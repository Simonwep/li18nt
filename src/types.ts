import {Conflicts, Duplicates, DuplicatesConfig, Indentation, PatternConfig, PatternMismatch, PrettifyOptions} from '@lib';

export type Mode = 'off' | 'warn' | 'error';
export type Li18ntOption<T> = [Mode, T?]

export interface Li18ntOptions {
    prettified: Indentation | PrettifyOptions;
    duplicates: boolean | DuplicatesConfig;
    conflicts: boolean;
    naming: PatternConfig;
}

export interface Li18ntResult {
    prettified: string[];
    duplicates: Duplicates[];
    conflicts: Conflicts;
    naming: PatternMismatch[][];
}

export type PartialLi18ntResult<T extends Partial<Li18ntOptions>> = {
    [P in keyof T]: T[P] extends undefined ?
        never : P extends keyof Li18ntResult ?
            Li18ntResult[P] : never
}

export type CLIRules = {
    prettified: Li18ntOption<PrettifyOptions>;
    duplicates: Li18ntOption<DuplicatesConfig>;
    conflicts: Li18ntOption<boolean>;
    naming: Li18ntOption<PatternConfig>;
}

export interface CLIOptions {
    rules: CLIRules;
    fix?: boolean;
    debug?: boolean;
    quiet?: boolean;
    config?: string;
    skipInvalid?: boolean;
    report?: boolean;
}

export interface SourceFile {
    content: JSONObject;
    source: string;
    filePath: string;
    name: string;
}

export interface CLIModuleArguments<Config> {
    files: SourceFile[];
    cmd: CLIOptions;
    rule: Li18ntOption<Config>;
}

export interface CLIModule<Config = undefined> {
    (args: CLIModuleArguments<Config>): boolean;
}

export type PropertyPath = (string | number)[];

export type JSONValue = string | number | boolean | null | JSONObject | JSONArray;
export type JSONArray = JSONValue[] | JSONObject[];

export interface JSONObject {
    [key: string]: JSONValue;
}
