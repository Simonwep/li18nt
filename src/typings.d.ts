declare const VERSION: string | undefined;
