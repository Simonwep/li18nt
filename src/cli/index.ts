import {CLIOptions, CLIRules, Mode} from '@types';
import {Command, program} from 'commander';
import {entry} from './entry';
import {warnLn} from './utils/log';
import {printReport} from './utils/printReport';
import {resolveConfiguration} from './utils/resolveConfiguration';

const version = typeof VERSION === 'undefined' ? 'unknown' : VERSION;

/* eslint-disable @typescript-eslint/no-explicit-any */
const processRule = (mode: Mode | [Mode, unknown]): [Mode, any] => {
    return typeof mode === 'string' ? [mode, undefined] : mode;
};

const undefinedOr = <A>(a: A | undefined, b: A | undefined) => {
    return typeof a !== 'undefined' ? a : b;
};

program
    .version(version, '-v, --version', 'Output the current version')
    .helpOption('-h, --help', 'Show this help text')
    .name('lint-i18n')
    .description('Lints your locales files, lint-i18n is an alias.')
    .usage('[files...] [options]')
    .arguments('[files...]')
    .option('-q, --quiet', 'Print only errors and warnings')
    .option('-d, --debug', 'Debug information')
    .option('-f, --fix', 'Tries to fix existing errors')
    .option('--config [path]', 'Configuration file path (it\'ll try to resolve one in the current directory)')
    .option('--skip-invalid', 'Skip invalid files without exiting')
    .option('--report', 'Print system information')
    .action((args: any, cmd: Command & CLIOptions) => {

        // Print report and exit immediately
        if (cmd.report) {
            return printReport(version);
        }

        // Try to resolve and load config file
        const options = resolveConfiguration(cmd);
        if (options) {

            // Override options
            cmd.quiet = undefinedOr(cmd.quiet, options.quiet);
            cmd.skipInvalid = undefinedOr(cmd.skipInvalid, options.skipInvalid);
            cmd.rules = options.rules || {};

            for (const [name, value] of Object.entries(cmd.rules)) {
                cmd.rules[name as keyof CLIRules] = processRule(value as Mode | [Mode, unknown]);
            }
        } else {
            return warnLn('Missing configuration file.');
        }

        return entry(args, cmd);
    })
    .parse();
