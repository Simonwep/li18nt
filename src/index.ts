import {conflicts} from '@tools/conflicts';
import {duplicates} from '@tools/duplicates';
import {prettify} from '@tools/prettify';
import {JSONObject, Li18ntOptions, Li18ntResult} from '@types';

// Export property path utility
export * from './utils/propertyPath';

// Current version
export const version = typeof VERSION !== 'undefined' ? VERSION : 'unknown';

/**
 * Lints objects using the given configuration.
 * @param conf
 * @param objects
 */
export const lint = (conf: Li18ntOptions, objects: JSONObject[]): Li18ntResult => {
    const res: Li18ntResult = {};

    if (conf.conflicts) {
        res.conflicts = conflicts(objects);
    }

    if (conf.duplicates) {
        const options = typeof conf.duplicates === 'boolean' ? {} : conf.duplicates;

        res.duplicates = [];
        for (const obj of objects) {
            res.duplicates.push(duplicates(obj, options));
        }
    }

    if (typeof conf.prettified !== 'undefined') {
        const options = typeof conf.prettified !== 'object' ? {
            indent: conf.prettified
        } : conf.prettified;

        res.prettified = [];
        for (const obj of objects) {
            res.prettified.push(prettify(obj, options));
        }
    }

    return res;
};
