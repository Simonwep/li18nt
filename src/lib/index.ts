export * from './conflicts';
export * from './duplicates';
export * from './pattern';
export * from './prettify';
export * from './linter';
