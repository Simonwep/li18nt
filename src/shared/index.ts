export * from './generateList';
export * from './keysFrom';
export * from './objects';
export * from './typeOfJsonValue';
