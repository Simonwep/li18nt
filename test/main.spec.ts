import {lint} from '../src';

describe('Main library', () => {

    it('Should properly lint the given objects', ()=>{
        expect(lint({
            prettify: 4,
            duplicates: true,
            conflicts: true
        }, [
            {
                'back': 'Zurück',
                'dashboard': 'Dashboard',
                'pages': {
                    'back': 'Zurück',
                    'dashboard': {
                        's': null,
                        'dashboard': 'Dashboard',
                        'footer': 'Fußbar'
                    }
                }
            },
            {
                'about': 'About',
                'back': 'Back',
                'dashboard': 'Dashboard',
                'pages': {
                    'back': 'Back',
                    'dashboard': {
                        's': [],
                        'dashboard': 'Dashboard',
                        'footer': [],
                        'panels': 'Panels'
                    },
                    'main': {
                        'dashboard': 'Dashboard'
                    }
                }
            }
        ])).toMatchSnapshot();
    });

    it('Should work with the example given in the README', ()=>{
        expect(lint({
            prettify: '\t',
            duplicates: true,
            conflicts: true
        }, [
            {a: 20, b: null, c: {x: 20}},
            {a: 50, b: 'Hello', c: {x: 100, y: 20}},
            {a: 'Five', b: 'Super', c: null}
        ]));
    });
});
