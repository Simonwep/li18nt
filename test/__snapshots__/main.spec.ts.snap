// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main library Should properly lint the given objects 1`] = `
Object {
  "conflicts": Array [
    Object {
      "conflicts": Array [
        Array [
          "pages",
          "dashboard",
          "s",
        ],
      ],
      "missing": Array [
        Array [
          "pages",
          "dashboard",
          "panels",
        ],
        Array [
          "pages",
          "main",
        ],
        Array [
          "about",
        ],
      ],
    },
    Object {
      "conflicts": Array [
        Array [
          "pages",
          "dashboard",
          "s",
        ],
      ],
      "missing": Array [],
    },
  ],
  "duplicates": Array [
    Map {
      "back" => Array [
        Array [
          "pages",
          "back",
        ],
      ],
      "dashboard" => Array [
        Array [
          "pages",
          "dashboard",
          "dashboard",
        ],
      ],
    },
    Map {
      "back" => Array [
        Array [
          "pages",
          "back",
        ],
      ],
      "dashboard" => Array [
        Array [
          "pages",
          "dashboard",
          "dashboard",
        ],
        Array [
          "pages",
          "main",
          "dashboard",
        ],
      ],
    },
  ],
  "prettified": Array [
    "{
    \\"back\\": \\"Zurück\\",
    \\"dashboard\\": \\"Dashboard\\",
    \\"pages\\": {
        \\"back\\": \\"Zurück\\",
        \\"dashboard\\": {
            \\"dashboard\\": \\"Dashboard\\",
            \\"footer\\": \\"Fußbar\\",
            \\"s\\": null
        }
    }
}",
    "{
    \\"about\\": \\"About\\",
    \\"back\\": \\"Back\\",
    \\"dashboard\\": \\"Dashboard\\",
    \\"pages\\": {
        \\"back\\": \\"Back\\",
        \\"dashboard\\": {
            \\"dashboard\\": \\"Dashboard\\",
            \\"footer\\": [],
            \\"panels\\": \\"Panels\\",
            \\"s\\": []
        },
        \\"main\\": {
            \\"dashboard\\": \\"Dashboard\\"
        }
    }
}",
  ],
}
`;
